### PR TITLE
editoast: refactor projects tests

### PR DESCRIPTION
**Pull request by Wadjetz**

### CODE REVIEW

**Review by hamz2a — APPROVED** [review]
LGTM, thanks !

File: editoast/src/modelsv2/projects.rs [hunk 1]
```diff
@@ -118,84 +118,101 @@ impl Project {
...
-            .model
+    async fn project_update() {
+        let db_pool = DbConnectionPoolV2::for_tests();
+        let created_project = create_project(&db_pool, "test_project_name").await;
+
+        let project_name = "update_name";
+        let project_budget = Some(1000);
+
+        // Patch a project
+        let update_project = created_project
```
Comment (leovalais) [inline editoast/src/modelsv2/projects.rs]: You could be using `.patch()` for this:

```rust
project.patch().name(...).budget(...).apply(conn).await
```

File: editoast/src/modelsv2/projects.rs [hunk 2]
```diff
@@ -118,84 +118,101 @@ impl Project {
...
+            .expect("Failed to retrieve project")
+            .expect("Project not found");
+
+        assert_eq!(&created_project, &project);
+
+        assert!(
+            Project::list(db_pool.get_ok().deref_mut(), SelectionSettings::new())
+                .await
+                .is_ok()
+        );
```
Comment (leovalais) [inline editoast/src/modelsv2/projects.rs]:
```suggestion
```

list is tested in another test below

File: editoast/src/modelsv2/projects.rs [hunk 3]
```diff
@@ -118,84 +118,101 @@ impl Project {
 #[cfg(test)]
 pub mod test {
     use rstest::rstest;
-    use std::sync::Arc;
+    use std::ops::DerefMut;
```
Comment (leovalais) [inline editoast/src/modelsv2/projects.rs:122]:
```suggestion
    use std::ops::DerefMut;
    use pretty_assertions::assert_eq;
```

It's a drop-in replacement for `assert_eq!`, but shows a colored diff on failure. It's already in the dependance tree so we might as well use it.

File: editoast/src/views/projects.rs [hunk 4]
```diff
@@ -301,102 +302,166 @@ impl From<ProjectPatchForm> for Changeset<Project> {
...
+            .creation_date(Utc::now().naive_utc())
+            .last_modification(Utc::now().naive_utc())
+            .tags(Tags::default())
+    }
+
+    async fn create_project(pool: &DbConnectionPoolV2) -> Project {
+        project_changeset()
+            .create(pool.get_ok().deref_mut())
+            .await
+            .expect("Failed to create project")
```
Comment (leovalais) [inline editoast/src/views/projects.rs]: This is mostly duplication of the functions in `modelsv2::project::test`. You could define a `modelsv2::fixtures` public module that could be reused in the views.

File: editoast/src/modelsv2/projects.rs [hunk 5]
```diff
@@ -118,84 +118,101 @@ impl Project {
...
+    pub fn project_changeset(name: &str) -> Changeset<Project> {
+        Project::changeset()
+            .name(name.to_owned())
+            .budget(Some(0))
+            .creation_date(Utc::now().naive_utc())
+            .last_modification(Utc::now().naive_utc())
+            .tags(Tags::default())
+    }
+
+    async fn create_project(db_pool: &DbConnectionPoolV2, name: &str) -> Project {
```
Comment (leovalais) [inline editoast/src/modelsv2/projects.rs]:
```suggestion
    async fn create_project(conn: &mut DbConnection, name: &str) -> Project {
```

File: editoast/src/views/projects.rs [hunk 6]
```diff
@@ -301,102 +302,166 @@ impl From<ProjectPatchForm> for Changeset<Project> {
...
+        let response: ProjectWithStudyCountList =
+            call_and_read_body_json(&app.service, request).await;
+
+        let project_retreived = response
+            .results
+            .iter()
+            .find(|p| p.project.id == created_project.id)
+            .unwrap();
+
+        assert_eq!(created_project.name, project_retreived.project.name);
```
Comment (leovalais) [inline editoast/src/views/projects.rs]:
```suggestion
        assert_eq!(created_project, project_retreived.project);
```

File: editoast/src/views/projects.rs [hunk 7]
```diff
@@ -301,102 +302,166 @@ impl From<ProjectPatchForm> for Changeset<Project> {
...
+
+        assert_eq!(response.project.name, updated_name);
+        assert_eq!(response.project.budget, Some(updated_budget));
+
+        let project = Project::retrieve(pool.get_ok().deref_mut(), response.project.id)
+            .await
+            .expect("Failed to retrieve project")
+            .expect("Project not found");
+
+        assert_eq!(project.name, updated_name);
```
Comment (leovalais) [inline editoast/src/views/projects.rs:450]:
```suggestion
        assert_eq!(project.name, updated_name);
        assert_eq!(project.budget, Some(updated_budget));
```

File: editoast/src/views/pagination.rs [hunk 8]
```diff
@@ -47,7 +47,7 @@ const DEFAULT_PAGE_SIZE: u64 = 25;
 ///
 /// We named the data field `result` to cope with the old pagination schema which
 /// enforced this name. For new paginated responses, the field name is up to your imagination :)
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, ToSchema)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
```
Comment (leovalais) [inline editoast/src/views/pagination.rs]:
```suggestion
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, ToSchema)]
#[cfg_attr(test, derive(Deserialize))]
```

same for other `Deserialize`

**Review by leovalais — APPROVED** [review]
Thanks for refactoring and improving these tests :)

I have a few comments but nothing really blocking (except for the `cfg_attr` one).

File: editoast/src/modelsv2/fixtures.rs [hunk 9]
```diff
@@ -0,0 +1,25 @@
+#[cfg(test)]
+pub mod tests {
```
Comment (leovalais) [inline editoast/src/modelsv2/fixtures.rs]: You don't need to make a test submodule, just `#[cfg(test)]` its inclusion in the source tree

File: editoast/src/modelsv2/fixtures.rs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    use chrono::Utc;
+
+    use crate::modelsv2::prelude::*;
+    use crate::modelsv2::DbConnection;
+    use crate::modelsv2::Project;
+    use crate::modelsv2::Tags;
+
+    pub fn test_project_changeset(name: &str) -> Changeset<Project> {
```
Comment (leovalais) [inline editoast/src/modelsv2/fixtures.rs]: Completely subjective comment + nitpick: I'd drop the `test` prefix as the function module already suggests we're in a test context (we have namespaces for a reason ^^)

File: editoast/src/modelsv2/mod.rs [hunk 11]
```diff
@@ -1,6 +1,7 @@
 pub mod database;
 pub mod documents;
 pub mod electrical_profiles;
+pub mod fixtures;
```
Comment (leovalais) [inline editoast/src/modelsv2/mod.rs:5]:
```suggestion
#[cfg(test)] pub mod fixtures;
```

File: editoast/src/modelsv2/projects.rs [hunk 12]
```diff
@@ -117,85 +117,86 @@ impl Project {
...
+
+        let project_name = "update_name";
+        let project_budget = Some(1000);
+
+        // Patch a project
+        created_project
             .clone()
-            .into_changeset()
-            .name(project.model.name.clone() + "_bis");
+            .patch()
```
Comment (leovalais) [inline editoast/src/modelsv2/projects.rs:164]: you don't need to clone() it (am I missing something? 🤔)

File: editoast/src/views/projects.rs [hunk 13]
```diff
@@ -120,7 +119,7 @@ async fn check_image_content(db_pool: Data<DbConnectionPool>, document_key: i64)
     Ok(())
 }
 
-#[derive(Debug, Clone, Serialize, ToSchema)]
+#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
```
Comment (leovalais) [inline editoast/src/views/projects.rs]: cfg_attr

**Review by leovalais — COMMENTED** [review]
A few more 😄

File: editoast/src/modelsv2/projects.rs (hunk 12 above)
Comment (Wadjetz) [reply editoast/src/modelsv2/projects.rs:164]: No, it wasn't necessary to clone. 👍

**Pull request merged**